numba/cuda/tests/cudapy/test_lineinfo.py

```diff
@@ -1,42 +1,207 @@
-        match = re_section_lineinfo.search(asm)
-        assertfn = self.assertIsNotNone if expect else self.assertIsNone
-        assertfn(match, msg=asm)
+        pat = (
+            r'\.file'                # .file directive beginning
+            r'\s*[0-9]*\s*'          # file number surrounded by whitespace
+            r'".*test_lineinfo.py"'  # filename in quotes, ignoring full path
+        )
+        match = re.compile(pat).search(ptx)
+        assertfn(match, msg=ptx)
+
+        # .loc directives should be present in the ptx
+        self._loc_directive_regex().search(ptx)
+        assertfn(match, msg=ptx)
+
+        # Debug info sections should not be present when only lineinfo is
+        # generated
+        pat = (
+            r'\.section\s*'  # .section directive beginning
+            r'\.debug_'      # Any section name beginning ".debug_"
+        )
+        match = re.compile(pat).search(ptx)
+        self.assertIsNone(match, msg=ptx)
 
     def test_no_lineinfo_in_asm(self):
         @cuda.jit(lineinfo=False)
         def foo(x):
             x[0] = 1
 
-        self._check(foo, sig=(types.int32[:],), expect=False)
+        self._check(foo, sig=(int32[:],), expect=False)
 
     def test_lineinfo_in_asm(self):
+        if not NVVM().is_nvvm70:
+            self.skipTest("lineinfo not generated for NVVM 3.4")
+
         @cuda.jit(lineinfo=True)
         def foo(x):
             x[0] = 1
 
-        self._check(foo, sig=(types.int32[:],), expect=True)
+        self._check(foo, sig=(int32[:],), expect=True)
+
+    def test_lineinfo_maintains_error_model(self):
+        sig = (float32[::1], float32[::1])
+
+        @cuda.jit(sig, lineinfo=True)
+        def divide_kernel(x, y):
+            x[0] /= y[0]
+
+        llvm = divide_kernel.inspect_llvm(sig)
+
+        # When the error model is Python, the device function returns 1 to
+        # signal an exception (e.g. divide by zero) has occurred. When the
+        # error model is the default NumPy one (as it should be when only
+        # lineinfo is enabled) the device function always returns 0.
+        self.assertNotIn('ret i32 1', llvm)
+
+    def test_no_lineinfo_in_device_function(self):
+        # Ensure that no lineinfo is generated in device functions by default.
+        @cuda.jit
+        def callee(x):
+            x[0] += 1
+
+        @cuda.jit
+        def caller(x):
+            x[0] = 1
+            callee(x)
+
+        sig = (int32[:],)
+        self._check(caller, sig=sig, expect=False)
+
+    def test_lineinfo_in_device_function(self):
+        if not NVVM().is_nvvm70:
+            self.skipTest("lineinfo not generated for NVVM 3.4")
+
+        # First we define a device function / kernel pair and run the usual
+        # checks on the generated LLVM and PTX.
+
+        @cuda.jit(lineinfo=True)
+        def callee(x):
+            x[0] += 1
+
+        @cuda.jit(lineinfo=True)
+        def caller(x):
+            x[0] = 1
+            callee(x)
```

Comment:
As noted OOB, I'm not sure it's not clear what the inheritance model is for the `lineinfo` flag. Does the device function need `lineinfo=True` if the calling kernel has `lineinfo=True` already. Further, what if the device function has `False` but the kernel has `True` (or vice versa).

Reply:
It is not necessary for the device function to have it (and `False` is the default). I think perhaps if this test had two variants, that check with `lineinfo=True` on the device function and without it, both expecting the exact same behaviour, that would cover the cases - does that sound suitable to you?

Reply:
Turns out I'm wrong, and you do need to add `lineinfo=True` to every function you want lineinfo for - it doesn't propagate.

Reply:
From OOB discussion: Changing this behaviour requires quite a large rework of the way flags and options work, and although I tried to fix that for this PR, it is quite a big scope bloat, so I'll have to follow up with another PR if we decide to change this soon. Some of the work towards that can be found in this branch (only provided for info, there's no need related to this PR to look at it): https://github.com/gmarkall/numba/tree/cuda-lineinfo-fixes-wip-20230216

Reply:
Thanks for capturing the discussion/outcomes over the handling of flags @gmarkall.